File: model-client/src/main/java/org/modelix/model/VersionMerger.java
package org.modelix.model;

import de.q60.mps.shadowmodels.runtime.model.persistent.IBranch;
import de.q60.mps.shadowmodels.runtime.model.persistent.IIdGenerator;
import de.q60.mps.shadowmodels.runtime.model.persistent.ITree;
import de.q60.mps.shadowmodels.runtime.model.persistent.IWriteTransaction;
import de.q60.mps.shadowmodels.runtime.model.persistent.PBranch;
import gnu.trove.set.TLongSet;
import gnu.trove.set.hash.TLongHashSet;
import org.apache.commons.lang3.mutable.MutableObject;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;
import org.modelix.StreamUtil;
import org.modelix.model.lazy.CLTree;
import org.modelix.model.lazy.CLVersion;
import org.modelix.model.lazy.IDeserializingKeyValueStore;
import org.modelix.model.operations.IAppliedOperation;
import org.modelix.model.operations.IOperation;
import org.modelix.model.persistent.CPVersion;

import java.util.ArrayList;
import java.util.HashSet;
import java.util.List;
import java.util.Objects;
import java.util.Optional;
import java.util.Set;
import java.util.stream.Collectors;

public class VersionMerger {

  private final Object mergeLock = new Object();

  private IDeserializingKeyValueStore storeCache;
  private IIdGenerator idGenerator;

  public VersionMerger(IDeserializingKeyValueStore storeCache, IIdGenerator idGenerator) {
    this.storeCache = storeCache;
    this.idGenerator = idGenerator;
  }

  @NotNull
  public CLVersion mergeChange(CLVersion lastMergedVersion, @NotNull CLVersion newVersion) {
    synchronized (mergeLock) {
      if (lastMergedVersion == null) {
        lastMergedVersion = newVersion;
        return newVersion;
      } else {
        if (Objects.equals(newVersion.getHash(), lastMergedVersion.getHash())) {
          return lastMergedVersion;
        }
        CLVersion merged = mergeHistory(lastMergedVersion.getHash(), newVersion.getHash());
        checkTreeIds(lastMergedVersion, newVersion);
        return merged;
      }
    }
  }

  public void checkTreeIds(CLVersion v1, CLVersion v2) {
    String id1 = v1.getTree().getId();
    String id2 = v2.getTree().getId();
    if (!(Objects.equals(id1, id2))) {
      throw new RuntimeException("Tree ID mismatch: " + id1 + " and " + id2);
    }
  }

  @NotNull
  protected CLVersion mergeHistory(String leftVersionHash, String rightVersionHash) {
    String commonBase = commonBaseVersion(leftVersionHash, rightVersionHash);

    final List<CLVersion> leftHistory = getHistory(leftVersionHash, commonBase);
    final List<CLVersion> rightHistory = getHistory(rightVersionHash, commonBase);

    final MutableObject<CLVersion> mergedVersion = new MutableObject<>();

    CLTree tree = getVersion(commonBase).getTree();
    if (tree == null) {
      tree = new CLTree(storeCache);
    }
    final IBranch branch = new PBranch(tree);

    if (rightHistory.isEmpty() || leftHistory.isEmpty()) {
      List<CLVersion> fastForwardHistory = leftHistory.isEmpty() ? rightHistory : leftHistory;
      int numOps = fastForwardHistory.stream().map(CLVersion::getNumberOfOperations).reduce(0, Math::max);
      if (numOps > 100) {
        return fastForwardHistory.get(0);
      }
      // A small number of changes may be faster to compute locally. 
    }

    branch.runWrite(() -> {
      IWriteTransaction t = branch.getWriteTransaction();

      List<IAppliedOperation> leftAppliedOps = new ArrayList<>();
      List<IAppliedOperation> rightAppliedOps = new ArrayList<>();
      TLongSet appliedVersionIds = new TLongHashSet();
      while (!leftHistory.isEmpty() || !rightHistory.isEmpty()) {
        boolean useLeft = rightHistory.isEmpty() || !leftHistory.isEmpty() && StreamUtil.last(leftHistory).getId() < StreamUtil.last(rightHistory).getId();
        CLVersion versionToApply = StreamUtil.removeLast(useLeft ? leftHistory : rightHistory);
        if (appliedVersionIds.contains(versionToApply.getId())) {
          continue;
        }
        appliedVersionIds.add(versionToApply.getId());
        final List<IOperation> oppositeAppliedOps = (useLeft ? rightAppliedOps : leftAppliedOps).stream()
                .map(IAppliedOperation::getOriginalOp)
                .collect(Collectors.toList());
        List<IOperation> operationsToApply = StreamUtil.toStream(versionToApply.getOperations())
                .map(it -> transformOperation(it, oppositeAppliedOps))
                .collect(Collectors.toList());
        for (IOperation op : operationsToApply) {
          IAppliedOperation appliedOp = op.apply(t);
          if (useLeft) {
            leftAppliedOps.add(appliedOp);
          } else {
            rightAppliedOps.add(appliedOp);
          }
        }

        mergedVersion.setValue(new CLVersion(
                versionToApply.getId(),
                versionToApply.getTime(),
                versionToApply.getAuthor(),
                ((CLTree) t.getTree()).getHash(),
                mergedVersion.getValue() != null
                        ? mergedVersion.getValue().getHash()
                        : versionToApply.getPreviousHash(),
                operationsToApply.toArray(new IOperation[operationsToApply.size()]),
                storeCache
        ));
      }
    });

    if (mergedVersion.getValue() == null) {
      throw new RuntimeException("Failed to merge " + leftVersionHash + " and " + rightVersionHash);
    }

    return mergedVersion.getValue();
  }

  protected IOperation transformOperation(IOperation opToTransform, Iterable<IOperation> previousOps) {
    IOperation result = opToTransform;
    for (IOperation previous : previousOps) {
      result = result.transform(previous);
    }
    return result;
  }

  /**
   * 
   * 
   * @param fromVersion The newest version
   * @param toVersionExclusive The oldest version
   * @return Newest version first
   */
  protected List<CLVersion> getHistory(@NotNull String fromVersion, @Nullable String toVersionExclusive) {
    List<CLVersion> history = new ArrayList<CLVersion>();
    if (Objects.equals(fromVersion, toVersionExclusive)) {
      return history;
    }
    CLVersion version = getVersion(fromVersion);
    while (true) {
      if (version == null) {
        break;
      }
      history.add(version);
      if (version.getPreviousHash() == null) {
        break;
      }
      if (Objects.equals(version.getPreviousHash(), toVersionExclusive)) {
        break;
      }
      version = version.getPreviousVersion();
    }
    return history;
  }

  protected String commonBaseVersion(String leftHash, String rightHash) {

    Set<String> leftVersions = new HashSet<>();
    Set<String> rightVersions = new HashSet<>();

    while (leftHash != null || rightHash != null) {
      if (leftHash != null) {
        leftVersions.add(leftHash);
      }
      if (rightHash != null) {
        rightVersions.add(rightHash);
      }

      if (leftHash != null) {
        if (rightVersions.contains(leftHash)) {
          return leftHash;
        }
      }
      if (rightHash != null) {
        if (leftVersions.contains(rightHash)) {
          return rightHash;
        }
      }

      if (leftHash != null) {
        leftHash = Optional.ofNullable(getVersion(leftHash)).map(CLVersion::getPreviousHash).orElse(null);
      }
      if (rightHash != null) {
        rightHash = Optional.ofNullable(getVersion(rightHash)).map(CLVersion::getPreviousHash).orElse(null);
      }
    }

    return null;
  }

  private @Nullable CLVersion getVersion(String hash) {
    if (hash == null) {
      return null;
    }
    return CLVersion.loadFromHash(hash, storeCache);
  }

  protected ITree getTree(CPVersion version) {
    if (version == null) {
      return null;
    }
    return new CLTree(version.treeHash, storeCache);
  }
}

File: model-client/src/main/kotlin/org/modelix/model/VersionMerger.kt
package org.modelix.model

import gnu.trove.set.TLongSet
import gnu.trove.set.hash.TLongHashSet
import org.apache.commons.lang3.mutable.MutableObject
import org.modelix.model.api.IBranch
import org.modelix.model.api.IIdGenerator
import org.modelix.model.api.ITree
import org.modelix.model.api.PBranch
import org.modelix.model.lazy.CLTree
import org.modelix.model.lazy.CLVersion
import org.modelix.model.lazy.IDeserializingKeyValueStore
import org.modelix.model.operations.IAppliedOperation
import org.modelix.model.operations.IOperation
import org.modelix.model.persistent.CPVersion
import org.modelix.model.util.StreamUtils
import java.util.Optional
import java.util.stream.Collectors
import kotlin.collections.ArrayList
import kotlin.collections.HashSet

class VersionMerger(private val storeCache: IDeserializingKeyValueStore, private val idGenerator: IIdGenerator) {
    private val mergeLock = Any()
    fun mergeChange(lastMergedVersion: CLVersion, newVersion: CLVersion): CLVersion {
        var lastMergedVersion = lastMergedVersion
        synchronized(mergeLock) {
            return if (lastMergedVersion == null) {
                lastMergedVersion = newVersion
                newVersion
            } else {
                if (newVersion.hash == lastMergedVersion.hash) {
                    return lastMergedVersion
                }
                val merged = mergeHistory(lastMergedVersion.hash, newVersion.hash)
                checkTreeIds(lastMergedVersion, newVersion)
                merged
            }
        }
    }

    fun checkTreeIds(v1: CLVersion, v2: CLVersion) {
        val id1 = v1.tree.id
        val id2 = v2.tree.id
        if (id1 != id2) {
            throw RuntimeException("Tree ID mismatch: $id1 and $id2")
        }
    }

    protected fun mergeHistory(leftVersionHash: String, rightVersionHash: String): CLVersion {
        val commonBase = commonBaseVersion(leftVersionHash, rightVersionHash)
        val leftHistory = getHistory(leftVersionHash, commonBase)
        val rightHistory = getHistory(rightVersionHash, commonBase)
        val mergedVersion = MutableObject<CLVersion>()
        var tree = getVersion(commonBase)!!.tree
        if (tree == null) {
            tree = CLTree(storeCache)
        }
        val branch: IBranch = PBranch(tree)
        if (rightHistory.isEmpty() || leftHistory.isEmpty()) {
            val fastForwardHistory = if (leftHistory.isEmpty()) rightHistory else leftHistory
            val numOps = fastForwardHistory.stream().map { obj: CLVersion -> obj.numberOfOperations }.reduce(0) { a: Int, b: Int -> Math.max(a, b) }
            if (numOps > 100) {
                return fastForwardHistory[0]
            }
            // A small number of changes may be faster to compute locally. 
        }
        branch.runWrite(
            Runnable {
                val t = branch.writeTransaction
                val leftAppliedOps: MutableList<IAppliedOperation> = ArrayList()
                val rightAppliedOps: MutableList<IAppliedOperation> = ArrayList()
                val appliedVersionIds: TLongSet = TLongHashSet()
                while (!leftHistory.isEmpty() || !rightHistory.isEmpty()) {
                    val useLeft = rightHistory.isEmpty() || !leftHistory.isEmpty() && StreamUtils.last(leftHistory)!!.id < StreamUtils.last(rightHistory)!!.id
                    val versionToApply = StreamUtils.removeLast(if (useLeft) leftHistory else rightHistory)
                    if (appliedVersionIds.contains(versionToApply!!.id)) {
                        continue
                    }
                    appliedVersionIds.add(versionToApply.id)
                    val oppositeAppliedOps = (if (useLeft) rightAppliedOps else leftAppliedOps).stream()
                        .map { obj: IAppliedOperation -> obj.originalOp }
                        .collect(Collectors.toList())
                    val operationsToApply = StreamUtils.toStream(versionToApply.operations)
                        .map { it: IOperation -> transformOperation(it, oppositeAppliedOps) }
                        .collect(Collectors.toList())
                    for (op in operationsToApply) {
                        val appliedOp = op.apply(t)
                        if (useLeft) {
                            leftAppliedOps.add(appliedOp!!)
                        } else {
                            rightAppliedOps.add(appliedOp!!)
                        }
                    }
                    mergedVersion.value = CLVersion(
                        versionToApply.id,
                        versionToApply.time,
                        versionToApply.author,
                        (t!!.tree as CLTree).hash,
                        if (mergedVersion.value != null) mergedVersion.value.hash else versionToApply.previousHash,
                        operationsToApply.toTypedArray(),
                        storeCache
                    )
                }
            }
        )
        if (mergedVersion.value == null) {
            throw RuntimeException("Failed to merge $leftVersionHash and $rightVersionHash")
        }
        return mergedVersion.value
    }

    protected fun transformOperation(opToTransform: IOperation, previousOps: Iterable<IOperation?>): IOperation {
        var result = opToTransform
        for (previous in previousOps) {
            result = result.transform(previous)!!
        }
        return result
    }

    /**
     *
     *
     * @param fromVersion The newest version
     * @param toVersionExclusive The oldest version
     * @return Newest version first
     */
    protected fun getHistory(fromVersion: String, toVersionExclusive: String?): MutableList<CLVersion> {
        val history: MutableList<CLVersion> = ArrayList()
        if (fromVersion == toVersionExclusive) {
            return history
        }
        var version = getVersion(fromVersion)
        while (true) {
            if (version == null) {
                break
            }
            history.add(version)
            if (version.previousHash == null) {
                break
            }
            if (version.previousHash == toVersionExclusive) {
                break
            }
            version = version.previousVersion
        }
        return history
    }

    protected fun commonBaseVersion(leftHash: String?, rightHash: String?): String? {
        var leftHash = leftHash
        var rightHash = rightHash
        val leftVersions: MutableSet<String> = HashSet()
        val rightVersions: MutableSet<String> = HashSet()
        while (leftHash != null || rightHash != null) {
            if (leftHash != null) {
                leftVersions.add(leftHash)
            }
            if (rightHash != null) {
                rightVersions.add(rightHash)
            }
            if (leftHash != null) {
                if (rightVersions.contains(leftHash)) {
                    return leftHash
                }
            }
            if (rightHash != null) {
                if (leftVersions.contains(rightHash)) {
                    return rightHash
                }
            }
            if (leftHash != null) {
                leftHash = Optional.ofNullable(getVersion(leftHash)).map { obj: CLVersion -> obj.previousHash }.orElse(null)
            }
            if (rightHash != null) {
                rightHash = Optional.ofNullable(getVersion(rightHash)).map { obj: CLVersion -> obj.previousHash }.orElse(null)
            }
        }
        return null
    }

    private fun getVersion(hash: String?): CLVersion? {
        return if (hash == null) {
            null
        } else CLVersion.loadFromHash(hash, storeCache)
    }

    protected fun getTree(version: CPVersion?): ITree? {
        return if (version == null) {
            null
        } else CLTree(version.treeHash, storeCache)
    }
}
